pkg/cache/groupcache.go

```diff
@@ -0,0 +1,222 @@
+	)
+
+	pool := groupcache.NewHTTPPoolOpts(conf.SelfURL, &groupcache.HTTPPoolOptions{
+		BasePath: basepath,
+	})
+
+	ticker := time.NewTicker(conf.DNSInterval)
+
+	go func() {
+		for {
+			if err := dnsGroupcacheProvider.Resolve(context.Background(), conf.Peers); err != nil {
+				level.Error(logger).Log("msg", "failed to resolve addresses for groupcache", "err", err)
+			} else {
+				pool.Set(dnsGroupcacheProvider.Addresses()...)
+			}
+
+			<-ticker.C
+		}
+	}()
+
+	r.Get(basepath, pool.ServeHTTP)
+
+	group := groupcache.NewGroup(conf.GroupcacheGroup, int64(conf.MaxSize), groupcache.GetterFunc(
+		func(ctx context.Context, id string, dest groupcache.Sink) error {
+			parsedData, err := cachekey.ParseBucketCacheKey(id)
+			if err != nil {
+				return err
+			}
+
+			switch parsedData.Verb {
+			case cachekey.AttributesVerb:
+				attrs, err := bucket.Attributes(ctx, parsedData.Name)
+				if err != nil {
+					return err
+				}
+
+				finalAttrs, err := json.Marshal(attrs)
+				if err != nil {
+					return err
+				}
+				err = dest.SetString(string(finalAttrs), time.Now().Add(5*time.Minute))
+				if err != nil {
+					return err
+				}
+			case cachekey.IterVerb:
+				// Not supported.
+
+				return nil
+			case cachekey.ContentVerb:
+				rc, err := bucket.Get(ctx, parsedData.Name)
+				if err != nil {
+					return err
+				}
+				defer runutil.CloseWithLogOnErr(logger, rc, "closing get")
+
+				b, err := ioutil.ReadAll(rc)
+				if err != nil {
+					return err
+				}
+
+				err = dest.SetBytes(b, time.Now().Add(5*time.Minute))
+				if err != nil {
+					return err
+				}
+			case cachekey.ExistsVerb:
+				exists, err := bucket.Exists(ctx, parsedData.Name)
+				if err != nil {
+					return err
+				}
+
+				err = dest.SetString(strconv.FormatBool(exists), time.Now().Add(5*time.Minute))
+				if err != nil {
+					return err
+				}
+			case cachekey.SubrangeVerb:
+				rc, err := bucket.GetRange(ctx, parsedData.Name, parsedData.Start, parsedData.End-parsedData.Start)
+				if err != nil {
+					return err
+				}
+				defer runutil.CloseWithLogOnErr(logger, rc, "closing get_range")
+
+				b, err := ioutil.ReadAll(rc)
+				if err != nil {
+					return err
+				}
+
+				err = dest.SetBytes(b, time.Now().Add(5*time.Minute))
+				if err != nil {
+					return err
+				}
+			}
+
+			return nil
+		},
+	))
+
+	return &Groupcache{
+		dns:    dnsGroupcacheProvider,
+		Pool:   pool,
+		group:  group,
+		logger: logger,
+	}, nil
+}
+
+func (c *Groupcache) Store(ctx context.Context, data map[string][]byte, ttl time.Duration) {
+	// Noop since cache is already filled during fetching.
+}
+
+func (c *Groupcache) Fetch(ctx context.Context, keys []string) map[string][]byte {
+	data := map[string][]byte{}
+
+	for _, k := range keys {
+		var keyData []byte
+
+		if err := c.group.Get(ctx, k, groupcache.AllocatingByteSliceSink(&keyData)); err != nil {
+			level.Error(c.logger).Log("msg", "failed fetching data from groupcache", "err", err, "key", k)
+			return nil
```

Comment:
Who wrote this code?!? :joy: :joy: maybe let's do `continue` here to return partial results if Get() fails?

Reply:
Let's do a git blame 😆